query-planner/src/model.rs

```diff
@@ -0,0 +1,48 @@
+use graphql_parser::query::{FragmentDefinition, SelectionSet, VariableDefinition};
+use indexmap::IndexSet;
+use std::collections::HashMap;
+use std::fmt;
+use std::fmt::{Display, Formatter};
+
+use crate::display;
+
+pub enum ResponsePathElement {
+    Field(String),
+    Idx(u32),
+}
+
+impl ToString for ResponsePathElement {
+    fn to_string(&self) -> String {
+        match self {
+            ResponsePathElement::Field(str) => str.to_string(),
+            ResponsePathElement::Idx(i) => i.to_string(),
+        }
+    }
+}
+
+pub struct QueryPlan<'a>(pub Option<PlanNode<'a>>);
+
+pub struct FetchNode<'a> {
+    pub service_name: String,
+    pub selection_set: SelectionSet<'a>,
+    pub variable_usages: Option<HashMap<String, VariableDefinition<'a>>>,
+    pub requires: Option<SelectionSet<'a>>,
+    pub internal_fragments: IndexSet<FragmentDefinition<'a>>,
+    pub source: String,
+}
+
+pub enum PlanNode<'a> {
```

Comment:
This is becoming a PITA to do now, as opposed to pushing this as an intermediate state. The reason is that the .ts code relies heavily on the `source` field to contain the actual operation, and there's not a lot of data that's used in the QueryPlan _execution_ from the ast nodes. I cleaned it up *a little*, but it's not worth the effort to do the rest now.

Specifically:
* from `FetchNode.variableUsages` we only get the variable names, so we don't need the `VariableDefinition` nodes, just strings. I made that change.
* `FetchNode.selectionSet` and `.internalFragments` are only used for _serialization_ (pretty printing), for _execution_, the data from the these fields is already incorporated into `source`.
* Finally, `FetchNode.requires` _is used_ for query **execution**, to build the variables field of the request with values fetched in a previous response. This whole thing is done in `executeSelectionSet` which is not easy to follow, I think @jbaxleyiii mentioned it builds a flat map of [path -> any] ? I don't remember. Anyway, since building this "right" now depends on how we're going to implement building the query plan, I'd rather stick with this intermediate world, so that we could iterate on wasm and print stuff easily.

Reply:
(I haven't given up on this fwiw, but might end up doing that if the effort doesn't prove valuable)

Reply:
@queerviolet The branch is updated!